Promote Order#all_adjustments to association

mbj opened the pull request:
This is part of a series of commits for performance improvements.
- Ability to eager load `#all_adjustments`
- Better reflection for tooling
- Better index usage as only `spree_adjustments.order_id` needs to get scanned
- Ability to add an FK to ensure DB side data integrity

This PR comes with a migration we used to fix integrity issues. that had been caused by bugs I already upstreamed fixes for before.

Other optimizations we did in our fork depend on this one.

mbj committed:
Promote Order#all_adjustments to association.

Fixes #6391

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://mountainroseherbs-fix-order-all-adjustments-association.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://mountainroseherbs-fix-order-all-adjustments-association.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

mbj commented on core/app/models/spree/order.rb:
```diff
@@ -70,6 +70,11 @@ def generate_number(options = {})
     has_many :variants, through: :line_items
     has_many :refunds, through: :payments
+    has_many :all_adjustments,
+      class_name: 'Spree::Adjustment',
```
@JDutil I do not think this recommendation makes sense here, as we than have to use explicit parenthesis about `has_many` DSL call.

```
has_many(
  :all_adjustments,
  class_name: ....,
  ...
)
```

Will not read that nice. Having the options in the same line violates the maximum line length, so I think the current format is the trade-off to do.

JDutil replied:
It could also be:
```
  has_many :all_adjustments,
                    class_name: 'Spree::Adjustment',
                    foreign_key: :order_id,
                    dependent: :destroy,
                    inverse_of: :order
```

mbj replied:
I do not like this style. But I want to get this patch in. So I changed to this recommendation.

mbj commented:
@JDutil This is ready for review.

It does not have a semantic change, but reduces code, and speeds up stuff a bit. So I request a backport to 2-3.

mbj commented:
@JDutil Ready for the next pass.

JDutil commented:
Thanks.  Usually we like to avoid migrations added to stable branches, but I think this is good to fix data integrity.

JDutil commented:
I've revert this for now since it's invalid mysql syntax:
```
Mysql2::Error: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'UPDATE spree_adjustments SET order_id = (SELECT order_id FROM spree_line_items W' at line 1: UPDATE spree_adjustments SET order_id = adjustable_id WHERE adjustable_type = 'Spree::Order' ; UPDATE spree_adjustments SET order_id = (SELECT order_id FROM spree_line_items WHERE spree_line_items.id = spree_adjustments.adjustable_id) WHERE adjustable_type = 'Spree::LineItem'/Users/JDutil/Code/spree/api/spec/dummy/db/migrate/20150527204053_fix_adjustment_order_id.spree.rb:5:in `change'
ActiveRecord::StatementInvalid: Mysql2::Error: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'UPDATE spree_adjustments SET order_id = (SELECT order_id FROM spree_line_items W' at line 1: UPDATE spree_adjustments SET order_id = adjustable_id WHERE adjustable_type = 'Spree::Order' ; UPDATE spree_adjustments SET order_id = (SELECT order_id FROM spree_line_items WHERE spree_line_items.id = spree_adjustments.adjustable_id) WHERE adjustable_type = 'Spree::LineItem'
/Users/JDutil/Code/spree/api/spec/dummy/db/migrate/20150527204053_fix_adjustment_order_id.spree.rb:5:in `change'
Mysql2::Error: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'UPDATE spree_adjustments SET order_id = (SELECT order_id FROM spree_line_items W' at line 1
/Users/JDutil/Code/spree/api/
```

Not sure why but it appears the nested select call isn't working as expected.

mbj commented:
> I've revert this for now since it's invalid mysql syntax:

I suspect the migration can easily be expressed in an AR loop, will be less performant but also work on mysql. I'll update the migration and keep the postgres compatible one in comments.

mbj commented:
@JDutil OT: You guys should consider to drop mysql for obvious reasons :P (dunno if you have paying clients for mysql support). My point here is that constraints on any level are helpful to create a value delivering product. Having to support anything results in achieveing nothing. Mysql is such a burden, especially on larger scale installations.

I'm realizing GH *is* on mysql, and this comment *is* stored in MySQL these moments, but I think GH people simply have more dev/operators to mitigate the effect.

An OSS project like spree might be better of dropping support for (a small fraction?) of your userbase, resulting in freeing resources to gain the larger fraction that is driven away by quality / performance issues.

Do you have any stats on how many of your users are on MySQL/PostgreSQL?

JDutil commented:
@mbj I think it can be done as AR loops as well, but obviously much prefer raw SQL.  I imagine we can fix it to work for both, but need to investigate it some more.  Just reverted soon as I noticed so that we could attempt to fix it before bug reports started coming in especially as it was ported to stable branches otherwise on master I could have let it live I suppose.

OT: I agree completely and from an official standpoint we have a long time ago, but unfortunately there are still a ton of people who use mysql most usually for legacy reasons however some are just unwilling to try new things so we should force them to switch in new apps.  We're actually experimenting with NoSQL databases as possible replacements altogether as the performance is above and beyond expectations way better we still need to work out some rough edges though.

Unfortunately there aren't great stats on usage for that I recently updated the alerts server, and the rails/spree versions people are requesting the alerts from aren't even that accurate IMO.  I'll have to look into adding DB stats to that as well though so that we can make a better decision.

mbj commented:
> but need to investigate it some more

So are you fixing this for mysql or I should? I want to significantly cut the diff we maintain, and this commit is the bottleneck for various other optimizations we'd like to upstream.

BTW If I where you I'd agree to suboptimal migration performance in MySQL using an AR based migration. As MySQL is not officially supported it should not cause improvements on master to be held back.

JDutil commented:
I've been wanting to fix it for MySQL with the tabs open for like a week but don't really have the time. I'm totally fine with making it a conditional migration that uses AR for non postgres adapters.

mbj commented:
> I'm totally fine with making it a conditional migration that uses AR for non postgres adapters.

K, I'll add that to the PR.

mbj commented:
@JDutil I added a MySQL code path for the data migration via AR usage. Its slower but will do the job for MySQL users. Also I added a comment explaining why there is an AR code path for MySQL users.

BTW, I'd love you port this back to at least 2-4-stable.

mbj commented:
@JDutil Ping. Anything else that prevents merge?

JDutil commented:
Thanks @mbj I've merged back to 2-3-stable.

mbj commented:
Cool. That enables me to upstream the more interesting changes.

jrochkind commented:
A bit late now, but I was surprised to find a change this big (refactor and db migration) released in a patch release, Spree 2.4.9. It's making it harder than I would like to upgrade my app currently using 2.4.6 to the latest (final) 2.4.x.

mbj commented:
> A bit late now, but I was surprised to find a change this big (refactor and db migration) released in a patch release, Spree 2.4.9. It's making it harder than I would like to upgrade my app currently using 2.4.6 to the latest (final) 2.4.x.

I do wounder what qualifies as big here, the semantic changes are minor.
